Run wasm tests on Helix

directhex opened the pull request:
Let's see what happens. Based on latest iOS PR.

premun committed:
Add XHarness Helix SDK support

premun committed:
Try to build mono libraries

premun committed:
Try to build mono libraries

premun committed:
Address some of Santiago's feedback from Jo's PR

premun committed:
Merge runtime.yml from master

premun committed:
Build iOSx64 libraries

premun committed:
Try to build mono product

premun committed:
Try to build mono product

premun committed:
Merge master

premun committed:
Do not build AllSubsets

premun committed:
Remove dead code

premun committed:
Revert: Do not built AllSubsets

premun committed:
Disable test archivation

directhex committed:
Improvements for local test running

directhex committed:
Merge branch 'prvysoky/use-helix-sdk' of github.com:premun/runtime into prvysoky/use-helix-sdk

premun committed:
Return dead code

directhex committed:
Reintroduce whitespace errors

None of these fixes are wrong per se (maybe the ApkBuilder changes?) but they complicate the PR by mixing unrelated changes

premun committed:
Bump Helix SDK and keep trailing slashes

premun committed:
Revert "Bump Helix SDK and keep trailing slashes"

This reverts commit d1a42261c34b5413f306ccaca37fed24294b8ac6.

premun committed:
Try building 2 app bundles

directhex committed:
Try a cleaner `.app` find in msbuild, from @akoeplinger

directhex committed:
Revert "Try a cleaner `.app` find in msbuild, from @akoeplinger"

This reverts commit 64da851be89f6f4dd00be7f804534aa3af15259d.

directhex committed:
Do not archiveTests on iOS, it causes tests to be included twice

directhex committed:
Revert "Revert "Try a cleaner `.app` find in msbuild, from @akoeplinger""

This reverts commit 506afd2c2b00800daf75b86f8ce48acb1c83ad84.

directhex committed:
Turn on all the tests on iOS

directhex committed:
Merge remote-tracking branch 'origin/master' into prvysoky/use-helix-sdk

directhex committed:
Remove Android changes, to keep scope of PR smaller

directhex committed:
Update global.json

Co-authored-by: Přemek Vysoký <premek.vysoky@microsoft.com>

directhex committed:
Update src/libraries/sendtohelix.proj

Co-authored-by: Přemek Vysoký <premek.vysoky@microsoft.com>

directhex committed:
Remove useless archiveTests=false

directhex committed:
Update src/libraries/sendtohelix.proj

Co-authored-by: Přemek Vysoký <premek.vysoky@microsoft.com>

premun committed:
Merge remote-tracking branch 'origin/master' into prvysoky/use-helix-sdk

premun committed:
Fix runtime-packs dir

safern committed:
Revert changes in yml to use new mono build workflow

safern committed:
Merge remote-tracking branch 'dotnet/master' into prvysoky/use-helix-sdk

safern committed:
Add iOS x64 testing to new collapsed build model

safern committed:
Just run System.Buffers.Tests to make testing faster

safern committed:
Set helixQueuesTemplate in the right place

premun committed:
Set XHarness version in Version.props

# Conflicts:
#	src/libraries/sendtohelix.proj

premun committed:
Roll back to XHarness that supports Xcode 11.4

safern committed:
Cleanup

premun committed:
Merge remote-tracking branch 'origin/master' into prvysoky/use-helix-sdk

# Conflicts:
#	eng/Versions.props

premun committed:
Add XHarness CLI to Version.Details.xml

directhex committed:
Merge remote-tracking branch 'origin/master' into wasm-helix-sdk

directhex committed:
Only test System.Buffers for now

directhex committed:
Try enabling testing for Wasm

directhex committed:
Move Wasm run to template which calls Helix

directhex committed:
Set a helix queue for wasm

safern commented on eng/pipelines/runtime.yml:
```diff
@@ -283,6 +281,33 @@ jobs:
+    buildConfig: Release
+    runtimeFlavor: mono
+    platforms:
+    - iOS_x64
```
I guess we should leave iOS where it was before merging for this PR. Maybe we should remove the iOS specific changes here.

safern commented on src/libraries/tests.proj:
```diff
@@ -21,7 +21,10 @@
     <ProjectReference Include="$(MSBuildThisFileDirectory)*\tests\**\*.Tests.csproj"
                       Exclude="@(ProjectExclusions)"
-                      Condition="'$(BuildAllConfigurations)' != 'true'" />
+                      Condition="'$(BuildAllConfigurations)' != 'true' and ('$(TargetsMobile)' != 'true' and '$(TargetsBrowser)' != 'true')" />
```
Let's leave this as it was as part of this PR. We can adjust later on the iOS/Android PRs.

directhex committed:
Ensure we archive tests on WebAssembly

directhex committed:
Be careful about using TargetsMobile, which is true on Wasm(?)

safern commented on src/libraries/sendtohelix.proj:
```diff
@@ -89,7 +89,7 @@
           Inputs="@(_RuntimeInputs);@(TestArchiveRuntimeDependency)"
           Outputs="$(TestArchiveRuntimeFile)"
-          Condition="'$(BuildAllConfigurations)' != 'true' and '$(TargetsMobile)' != 'true'">
+          Condition="'$(BuildAllConfigurations)' != 'true' and '$(TargetOS)' != 'iOS' and '$(TargetOS)' != 'tvOS' and '$(TargetOS)' != 'Android'">
```
As a suggestion we could introduce a property here in sendtohelix.proj to know if we're targeting any of these and share it for all the conditions.

premun replied:
I always thought it's `$(TargetsMobile)`?

safern replied:
TargetsMobile includes wasm as part of it so this si trying to do `TargetsMobile!=true or TargetOS == Browser` basically

akoeplinger replied:
yeah I think we can go with `TargetsMobile != true or TargetOS == Browser` like we do elsewhere

directhex committed:
Override artifact directory on wasm

directhex committed:
Let's install dotnet and xharness for wasm

directhex committed:
fixup

directhex committed:
HELIX_CORRELATION_PAYLOAD is readonly in docker

safern commented on eng/testing/tests.targets:
```diff
@@ -26,11 +26,15 @@
     <Error Condition="'$(TestArchiveTestsDir)' == ''" Text="TestArchiveTestsDir property to archive the test folder must be set." />
 
+    <PropertyGroup Condition="'$(TargetOS)' == 'Browser'">
+      <OutDir>$(OutDir)/AppBundle</OutDir>
```
I think we should use another property for this... Something like:

```xml
<PropertyGroup>
   <_dirToCompress>$(OutDir)</_dirToCompress>
   <_dirToCompress Condition="'$(TargetOS)' == 'Browser'">$(OutDir)/AppBundle</_dirToCompress>
</PropertyGroup>
```

and then change ZipDirectory to take _dirToCompress as the SourceDirectory.

Also, we can move: https://github.com/dotnet/runtime/blob/d2f9adc5ffd50ff5df09090f77ae8364656dc37d/eng/testing/tests.mobile.targets#L4 into tests.targets so that we don't have to duplicate the AppBundle dir. 

Changing the `OutDir` property is not ideal.

directhex replied:
Sure, I was just testing my theory that this was why the wrong thing was happening. It's a bad fix as-is

safern replied:
Yeah.. just adding the comments as I see stuff so that we don't forget addressing on the final iteration... just make it work, no need to address these comments now 😄

directhex committed:
Handle xharness directory

directhex committed:
Fiddle Helix launcher to run test suite

safern committed:
Cleanup RunTests script generation for wasm

safern committed:
Give runtime.js executable permisions

safern committed:
Fix ArchiveTest target to run after app bundle is created

premun committed:
Pre-install XHarness CLI using the Helix SDK

premun committed:
Pre-install XHarness CLI using the Helix SDK

premun committed:
Revert "Pre-install XHarness CLI using the Helix SDK"

This reverts commit a7b7b8f845710a75d96f5a7d154a690f5878aaee.

premun commented:
I tried to use the Helix SDK to preinstall XHarness but because it wants to install it into the correlation payload directory (as Matt has suggested earlier), [it fails](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-37752-merge-520227851947498a94/System.Buffers.Tests/console.544b8859.log?sv=2019-02-02&se=2020-07-05T09%3A22%3A20Z&sr=c&sp=rl&sig=wOgVVSW2VjXqkCSpQrTAmY6Gjb9xi%2BPD1sftLwoOg1w%3D):
```
Failed to install tool package 'microsoft.dotnet.xharness.cli': Read-only file system
```

directhex commented:
@premun ok, that doesn't hugely surprise me, that docker readonly thing is why I put xharness in `$HELIX_WORKITEM_PAYLOAD` instead of `$HELIX_CORRELATION_PAYLOAD`

premun commented:
@directhex we could temporarily change the variable technically, but seems hacky.. I wanted to put it in `$HELIX_WORKITEM_PAYLOAD` originally as well but Matt says it could be reused between work items.

directhex commented:
We probably need to loop Matt in to that discussion, then

directhex committed:
Merge remote-tracking branch 'origin/master' into wasm-helix-sdk

directhex committed:
Turn on all tests on wasm

directhex committed:
Skip test suite which hangs

directhex committed:
Disable test suites which don't emit result XML

directhex committed:
Skip all tests which don't pass yet

directhex committed:
Increase timeout on global-build-job

It can take more than an hour for a full build of runtime+libs+tests and helix run of tests

directhex committed:
Normalise on Windows path seperators

directhex committed:
Merge branch 'wasm-helix-sdk' of github.com:directhex/runtime into wasm-helix-sdk

directhex committed:
Merge remote-tracking branch 'origin/master' into wasm-helix-sdk

directhex committed:
Remove another test suite which doesn't pass any more

directhex committed:
Merge remote-tracking branch 'origin/master' into wasm-helix-sdk

directhex committed:
Merge remote-tracking branch 'origin/master' into wasm-helix-sdk

directhex committed:
Try using HelixPreCommand Include, not a big HelixPreCommands string

directhex committed:
ItemGroup not PropertyGroup

directhex committed:
Typo

directhex committed:
More typos

directhex committed:
Merge remote-tracking branch 'origin/master' into wasm-helix-sdk

directhex committed:
Remove iOS changes, and prepare for un-drafting

directhex committed:
More fixups

directhex committed:
Merge remote-tracking branch 'origin/master' into wasm-helix-sdk

directhex committed:
Merge remote-tracking branch 'origin/master' into wasm-helix-sdk

directhex committed:
Move directory zip definition to the right place

directhex committed:
Put dependency back how it was (gotta figure out why it's bad though)

akoeplinger commented on eng/testing/WasmRunnerTemplate.sh:
```diff
@@ -1,19 +1,20 @@
 _exitCode=$?
 
-echo "Xharness artifacts: $XHARNESS_OUT"
+echo "Xharness artifacts: `ls -lh $XHARNESS_OUT`"
```
why do we need this? I actually want to get the directory and not which files, or at least not in a format like this:

```
  XHarness exit code: 1001
  Xharness artifacts: total 64
  -rw-r--r--  1 alexander  staff    29K Jun 18 18:32 testResults.xml
```

premun replied:
Tbh, maybe having `set -x` at the beginning would do the trick much better?

akoeplinger replied:
@premun how would that help? this is mostly for local runs where a human looks at the output and I want to Ctrl/Cmd+Click on the path in my terminal to open the folder/file.

akoeplinger committed:
Fix circular dependency issue

akoeplinger commented on eng/testing/tests.targets:
```diff
@@ -31,8 +31,14 @@
+    <PropertyGroup>
+      <BundleDir>$([MSBuild]::NormalizeDirectory('$(OutDir)', 'AppBundle'))</BundleDir>
+      <_zipSourceDirectory>$(OutDir)</_zipSourceDirectory>
+      <_zipSourceDirectory Condition="'$(TargetOS)' == 'Browser'">$(BundleDir)</_zipSourceDirectory>
```
nit: make this `_ZipSourceDirectory `

safern commented on eng/testing/tests.targets:
```diff
@@ -28,10 +28,16 @@
           Condition="'$(ArchiveTests)' == 'true' and '$(IgnoreForCI)' != 'true' and
                      ('$(TargetsMobile)' != 'true' or '$(TargetOS)' == 'Browser')"
-          AfterTargets="Build"
+          AfterTargets="PrepareForRun"
```
The reason why I changed this was because this will run twice and the second time it runs it will overwrite the .zip with the wrong .zip.

The reason for that is because we have an inner and outerbuild.

safern replied:
Actually it was because: the ArchiveTests target has an `AfterTargets=PrepareForRun` which is executed inside `Build` target... our publish target has an `AfterTargets=Build` because `Publish` depends on Build; so for browser, we were essentially zipping up the folder before publishing the app and building the bundle, resulting in an empty zip.

akoeplinger replied:
Ah I see, that makes sense :)

safern replied:
I fixed it to not have the circular dependency now 😄

safern committed:
Sequence ArchiveTests correctly for browser

safern commented on eng/pipelines/runtime.yml:
```diff
@@ -283,6 +282,32 @@ jobs:
+      extraStepsTemplate: /eng/pipelines/libraries/helix.yml
+      extraStepsParameters:
+        creator: dotnet-bot
+        testRunNamePrefixSuffix: Mono_$(_BuildConfig)
```
Maybe we should add support to condition the extra steps... I can imagine the tests being executed when there is an installer only change. I think it would only make sense to run wasm tests if libraries or mono tests change, but build if libraries, mono or installer changes.

akoeplinger replied:
> I think it would only make sense to run wasm tests if libraries or mono tests change, but build if libraries, mono or installer changes.

Not sure if I understood you correctly here, but we do need to run the wasm tests if the libraries or mono implementation changes, not just the tests.

safern replied:
What I mean is run the tests if mono or libraries change only, don’t run the tests if only installer is changed. We can do that in a follow up PR though

directhex commented on src/libraries/tests.proj:
```diff
@@ -20,6 +20,137 @@
+    <ProjectExclusions Include="$(MSBuildThisFileDirectory)System.IO\tests\System.IO.Tests.csproj" />
+    <ProjectExclusions Include="$(MSBuildThisFileDirectory)System.Linq.Parallel\tests\System.Linq.Parallel.Tests.csproj" />
+    <ProjectExclusions Include="$(MSBuildThisFileDirectory)System.Net.Http.Json.Functional\tests\System.Net.Http.Json.Functional.Tests.csproj" />
+    <ProjectExclusions Include="$(MSBuildThisFileDirectory)System.Net.Mail.Functional\tests\System.Net.Mail.Functional.Tests.csproj" />
```
Some of the tests fail, this is a quick way to get the PR into a mergeable state, since annotating every failing test is a bigger job & we want to start testing ASAP

safern replied:
Ah I see. Well this will mess up the local execution... so maybe we should condition this project exclusions if `'$(ContinuousIntegrationBuild)' == 'true'`. 

Did you do a run after Mitches change and my change to fix the VFS?

directhex replied:
Literally the next thing I do today!

directhex replied:
Also, not a bad idea gating on ContinuousIntegrationBuild

directhex replied:
I might gate it behind a new property devs can set on the command line, so by default the same tests run on CI as locally but people trying to work on those have only minimal changes to make

safern replied:
I don't expect devs to run "all" tests locally, just people like us trying to get them in a clean state with bulk fixes. So I'm fine either way.

safern replied:
Also, FWIW, if we wanted to protect the build of these test projects in WASM, we could make them just not run in helix at all. That can be achieve by adding a `IgnoreForCI` property to true in each test project when `TargetOS == Browser`.

directhex committed:
Deal with nits from Alex

directhex committed:
Refresh list of failing tests

directhex committed:
Merge remote-tracking branch 'origin/master' into wasm-helix-sdk

directhex committed:
Define msbuildScript in global-build jobs (i.e. wasm)

safern committed:
Add cleanup in helix.yml missed on merge with master

safern committed:
Don't compress teshost for Browser

premun commented on eng/pipelines/common/global-build-job.yml:
```diff
@@ -8,10 +8,12 @@ parameters:
   crossrootfsDir: ''
   variables: {}
-  timeoutInMinutes: ''
+  timeoutInMinutes: 120
```
I am not sure this is part of the WIP, but you're changing the default for all legs, I guess?

safern replied:
Yeah I think we should intentionally pass it down as 120 only for legs that run tests from runtime.yml as a jobParameter.

directhex replied:
I hit a 60 minute whole-job timeout during one of my runs, this was a quick hack to not run into that.

directhex replied:
https://dev.azure.com/dnceng/public/_build/results?buildId=688824&view=logs&jobId=108d2c4a-8a62-5a58-8dad-8e1042acc93c

safern replied:
Yeah it makes sense for jobs that run tests, so we should just override the timeout in the job declaration in runtime.yml

premun commented on src/libraries/sendtohelixhelp.proj:
```diff
@@ -62,6 +62,17 @@
+    <DotNetCliVersion>3.1.201</DotNetCliVersion>
+  </PropertyGroup>
+  <ItemGroup Condition="'$(TargetOS)' == 'Browser'">
+    <HelixPreCommand Include="${HELIX_CORRELATION_PAYLOAD}/dotnet/dotnet tool install --tool-path ${HELIX_WORKITEM_PAYLOAD}/xharness-cli --version $(MicrosoftDotNetXHarnessCLIVersion) --add-source https://dnceng.pkgs.visualstudio.com/public/_packaging/dotnet-eng/nuget/v3/index.json Microsoft.DotNet.XHarness.CLI" />
+    <HelixPreCommand Include="export DOTNET_ROOT=${HELIX_CORRELATION_PAYLOAD}/dotnet" />
+    <HelixPreCommand Include="export PATH=${HELIX_WORKITEM_PAYLOAD}/xharness-cli:$PATH" />
```
This could now theoretically be replaced with setting a property
```xml
    <IncludeXHarnessCli>true</IncludeXHarnessCli>
```

And the section with `IncludeDotNetCli` can either be removed completely, or just the `IncludeDotNetCli` prop has to stay.

And once we improve the Helix SDK to pre-install the .NET runtime only and pack the tool as correlation payload, it will happen automatically here

directhex replied:
We can do this when we're not using Docker to run, which we can do when v8 is available on a non-docker queue

safern replied:
Can’t we just fix the docker image to work with the default HelixSDK switches? It is handy to get a repro environment when tests are run in a docker image. That way I can just run them in a container using the same image locally and I get the same behavior

akoeplinger commented on src/libraries/tests.proj:
```diff
@@ -20,6 +20,124 @@
+    <!-- Mysteriously won't build on Wasm -->
+    <ProjectExclusions Include="$(MSBuildThisFileDirectory)System.Runtime.Serialization.Formatters\tests\System.Runtime.Serialization.Formatters.Tests.csproj" />
+    <ProjectExclusions Include="$(MSBuildThisFileDirectory)System.Data.Common\tests\System.Data.Common.Tests.csproj" />
+    <ProjectExclusions Include="$(MSBuildThisFileDirectory)Microsoft.Extensions.DependencyModel\tests\Microsoft.Extensions.DependencyModel.Tests.csproj" />
```
You can enable these again, they build now.

MaximLipnin replied:
Perhaps, not to enable but move to other comment section, they still have failures

MaximLipnin commented on src/libraries/tests.proj:
```diff
@@ -20,6 +20,124 @@
+    <ProjectExclusions Include="$(MSBuildThisFileDirectory)System.ComponentModel.Annotations\tests\System.ComponentModel.Annotations.Tests.csproj" />
+    <ProjectExclusions Include="$(MSBuildThisFileDirectory)System.ComponentModel.Primitives\tests\System.ComponentModel.Primitives.Tests.csproj" />
+    <ProjectExclusions Include="$(MSBuildThisFileDirectory)System.ComponentModel.TypeConverter\tests\System.ComponentModel.TypeConverter.Tests.csproj" />
+    <ProjectExclusions Include="$(MSBuildThisFileDirectory)System.Console\tests\System.Console.Tests.csproj" />
```
It should pass now

directhex replied:
```
        WASM: [FAIL] ReadKey.CapsLock_GetUnix_ThrowsPlatformNotSupportedException
        WASM: [FAIL] ReadKey.NumberLock_GetUnix_ThrowsPlatformNotSupportedException
        WASM: [FAIL] WindowAndCursorProps.WindowLeft_GetUnix_ReturnsZero
        WASM: [FAIL] WindowAndCursorProps.WindowTop_GetUnix_ReturnsZero
```

MaximLipnin replied:
On the latest master I have:
```
WASM: Tests run: 85, Errors: 0, Failures: 0, Skipped: 9. Time: 0.5798649s
```

directhex replied:
Dunno what to tell you man, System.Console just isn't giving me that on my Macbook

`        WASM: Tests run: 85, Errors: 0, Failures: 4, Skipped: 9. Time: 0.390183s`

akoeplinger replied:
I have 0 failures on master too. Note that you need to be on master from ~4 hours ago after b2890c6b3b10185885fe4350944c7fc72d88ec73 was merged.

directhex replied:
I'll try a git clean build? 🤷‍♂️

directhex replied:
OK, why did that help? ._.

directhex committed:
Merge remote-tracking branch 'origin/master' into wasm-helix-sdk

directhex committed:
non-building tests now build

directhex committed:
Use Helix SDK to install xharness

directhex committed:
Merge remote-tracking branch 'origin/master' into wasm-helix-sdk

directhex committed:
Revert "Use Helix SDK to install xharness"

This reverts commit 42958debf69ba88c78274f1bf4bba917294f19c2.

directhex committed:
Enable System.Console

directhex committed:
Skip another suite, which maybe only started being included via dependencies?

directhex committed:
Move timeoutInMinutes to the right place

safern commented:
Taking a final look.

akoeplinger committed:
Merge branch 'master' into wasm-helix-sdk

akoeplinger committed:
Reenable test projects that pass now

premun commented on src/libraries/sendtohelixhelp.proj:
```diff
@@ -62,6 +62,17 @@
+  <ItemGroup Condition="'$(TargetOS)' == 'Browser'">
+    <HelixPreCommand Include="${HELIX_CORRELATION_PAYLOAD}/dotnet/dotnet tool install --tool-path ${HELIX_WORKITEM_PAYLOAD}/xharness-cli --version $(MicrosoftDotNetXHarnessCLIVersion) --add-source https://dnceng.pkgs.visualstudio.com/public/_packaging/dotnet-eng/nuget/v3/index.json Microsoft.DotNet.XHarness.CLI" />
+    <HelixPreCommand Include="export DOTNET_ROOT=${HELIX_CORRELATION_PAYLOAD}/dotnet" />
+    <HelixPreCommand Include="export PATH=${HELIX_WORKITEM_PAYLOAD}/xharness-cli:$PATH" />
```
This will improve the XHarness logs:
```suggestion
    <HelixPreCommand Include="export PATH=${HELIX_WORKITEM_PAYLOAD}/xharness-cli:$PATH" />
    <HelixPreCommands>$(HelixPreCommands) export XHARNESS_DISABLE_COLORED_OUTPUT=true;</HelixPreCommands>
    <HelixPreCommands>$(HelixPreCommands) export XHARNESS_LOG_WITH_TIMESTAMPS=true;</HelixPreCommands>
```

However, it would be better to add these commands only for CI, not local runs. Not sure how to do that?

premun replied:
Oh, sorry @directhex.. this should be the ItemGroup now. I forgot you changed that in runtime

directhex committed:
Update src/libraries/sendtohelixhelp.proj

Co-authored-by: Přemek Vysoký <premek.vysoky@microsoft.com>

directhex committed:
Fixup for Premek's change

directhex committed:
Merge remote-tracking branch 'origin/master' into wasm-helix-sdk

directhex committed:
ActiveIssue on Crypto test in Format.Cbor

akoeplinger committed:
Another ActiveIssue on Cbor Crypto test

akoeplinger committed:
Merge remote-tracking branch 'upstream/master' into wasm-helix-sdk

directhex committed:
Merge remote-tracking branch 'origin/master' into wasm-helix-sdk

akoeplinger committed:
Merge remote-tracking branch 'upstream/master' into wasm-helix-sdk

akoeplinger committed:
Enable three more testsuites that pass now

akoeplinger committed:
Reenable System.Reflection.Emit.ILGeneration

directhex committed:
Revert "Revert "Use Helix SDK to install xharness""

This reverts commit afaffb8fd4f802459dd00dab00ab3d59c27ce707.

directhex committed:
Run wasm tests in bare Ubuntu 20.04 queue, not in Docker

akoeplinger committed:
Merge remote-tracking branch 'upstream/master' into wasm-helix-sdk

akoeplinger committed:
Reenable two more test suites

akoeplinger committed:
Merge remote-tracking branch 'upstream/master' into wasm-helix-sdk

akoeplinger committed:
Reenable three more test suites

directhex committed:
Merge branch 'wasm-helix-sdk' of github.com:directhex/runtime into wasm-helix-sdk

safern commented:
Wuhu 🎉
